rfcs/rfc-colin-states-mapping-to-lear.md

```diff
@@ -0,0 +1,240 @@
+
+#### COLIN Database Structure for Business States
+![colin_states.png](rfc-colin-states-mapping-to-lear/colin_states.png)
+
+#### Current Corp State Data in CPRD
+_`ACT/HIS` column is `CORP_OP_STATE` table._
+_`Type` column is `CORP_STATE` table._
+![colin_corp_state_stats.png](rfc-colin-states-mapping-to-lear/colin_corp_state_stats.png)
+
+
+# Basic design
+
+The COLIN business states can be mapped 1-1 to the business states(`businesses.state`) in LEAR.  Only `ACTIVE` and `HISTORICAL` are valid business states.
+
+The `CORP_STATE` table in COLIN will not be used. It was essentially used track the latest filing a business had filed and not an actual state.  This functionality will be achieved in LEAR by using a combination of the filing type, dynamic calculations (warning structures or business properties like good standing) or using the filing type that transitioned a business into its current state.
+
+Compliance is informational only and does not block any registry actions but must be noted to viewers of the Registry.
+
+Good standing has various statuses and can lead to a business state transition.
+
+`Liquidation` is not a business state.  It will consist of a series of flow statuses similar to how _good standing_ works, eventually transitioning a business into a `historical` state.
+
+Flows like `good standing`, `liquidation`, `involuntary dissolution` will make use of existing business warnings structure to keep track of flow statuses.
+
+A periodic job(s) will run to analyze flow statuses and transition business state and do things like initiate a dissolution filing where required.
+
+# Detailed design
+
+Business GET endpoint will return an `allowableActions` property containing allowed "filing" actions.  The allowable filings will be determined using a combination of the allowable filing transitions map, [logic in the FE](https://github.com/bcgov/business-filings-ui/blob/main/src/mixins/allowable-actions-mixin.ts) that needs to be implemented in the BE and any other required logic.
+
+The `allowableActions` property can also be extended in the future to support things like "admin" actions.
+
+_Examples of what this might look like but will need to represent sub-types of filings where applicable_
+``` http response
+
+GET https://legal-api-dev.apps.silver.devops.gov.bc.ca/api/v2/businesses/BC0001199
+
+{
+    "business": {
+        "legalName": "XYZ TEST CORP.",
+        "legalType": "BC",
+        "state": "HISTORICAL",
+        "allowableActions": {
+            "admin": ["addDetailComment", "addStaffComment", "viewChangeCompanyInfo"]
+            "filing": {
+                "types": ["restoration", "putBackOn"],
+                "subTypes": {
+                   "restoration": ["fullRestoration", "limitedRestoration"]
+                }
+            }
+        }
```

Comment:
This looks good and I can definitely use it as is.  If possible, it would be even better if allowable actions were objects instead of strings that included a URI and name attributes.  If I were an outside consumer of the API, an allowable object tells me not only what I can do, but what URL to use to submit the action.

Reply:
Something like this?  Just a note that all the filings will pretty much have the same action url of `https://legal-api-dev.apps.silver.devops.gov.bc.ca/api/v2/businesses/BC0871274/filings` essentially.  It still provides a level of clarity to the api consumer though.

``` json
{
    "business": {
        "legalName": "XYZ TEST CORP.",
        "legalType": "BC",
        "state": "HISTORICAL",
        "allowableActions": {
            "admin": [
                {
                    "name": "addDetailComment",
                    "url": "https://legal-api-dev.apps.silver.devops.gov.bc.ca/api/v2/businesses/BC0871274/filings/{filing_id}/comments",
                    "description": "Add staff detail comment to a filing."
                },
                {
                    "name": "addStaffComment",
                    "url": "https://legal-api-dev.apps.silver.devops.gov.bc.ca/api/v2/businesses/BC0871274/comments",
                    "description": "Add staff comment to a filing."
                }
            ],
            "filing": [
                {
                    "name": "restoration",
                    "subTypes": [
                        {
                            "name": "fullRestoration",
                            "url": "https://legal-api-dev.apps.silver.devops.gov.bc.ca/api/v2/businesses/BC0871274/filings",
                            "description": "File full restoration filing."
                        },
                        {
                            "name": "limitedRestoration",
                            "url": "https://legal-api-dev.apps.silver.devops.gov.bc.ca/api/v2/businesses/BC0871274/filings",
                            "description": "File limited restoration filing."
                        }
                    ]
                },
                {
                    "name": "putBackOn",
                    "url": "https://legal-api-dev.apps.silver.devops.gov.bc.ca/api/v2/businesses/BC0871274/filings",
                    "description": "File put back on filing."
                }
            ]
        }
    }
}

```

Reply:
This is nice but IMO maybe too much, especially since we'd like to keep the `filings` response concise (since we don't page the data and there could be a lot of data).

@thorwolpert ?

Reply:
I'd be tempted to make the URL higher in the block level, and then all the filings that can be submitted.
for each filing, what info do we want to blast back? looking at GitHub and Google for inspiration they have some pretty verbose responses.
We've had requests for the feecodes, as users of the responses want to know what a filing costs. Without the feecodes / sku they can't lookup the price in pay.
Is it worth adding the common displayName?
If we're not adding that to the response, where should we provide that?
If we're worried about the volume of data, we could move this part of the response to an _actions_ sub-endpoint, or allow for a ?verbose=minimum flag to allow the caller to reduce.
I do have a soft spot for the more HATEOS style that we've leaned towards elsewhere. 
I know @argush3 is a fan too, so others are going to need to provide an alternate argument.